test/integration/daemonset/daemonset_test.go

```diff
@@ -498,6 +499,54 @@ func TestSimpleDaemonSetLaunchesPods(t *testing.T) {
 	})
 }
 
+func TestSimpleDaemonSetRestartsPods(t *testing.T) {
+	for _, podPhase := range []v1.PodPhase{v1.PodSucceeded, v1.PodFailed} {
+		t.Run(fmt.Sprintf("%s_%v", t.Name(), podPhase), func(tt *testing.T) {
+			forEachStrategy(tt, func(t *testing.T, strategy *apps.DaemonSetUpdateStrategy) {
+				ctx, closeFn, dc, informers, clientset := setup(t)
+				defer closeFn()
+				ns := framework.CreateNamespaceOrDie(clientset, "simple-daemonset-restart-test", t)
+				defer framework.DeleteNamespaceOrDie(clientset, ns, t)
+
+				dsClient := clientset.AppsV1().DaemonSets(ns.Name)
+				podClient := clientset.CoreV1().Pods(ns.Name)
+				nodeClient := clientset.CoreV1().Nodes()
+				podInformer := informers.Core().V1().Pods().Informer()
+
+				informers.Start(ctx.Done())
+				go dc.Run(ctx, 2)
+
+				ds := newDaemonSet("restart-terminal-pods", ns.Name)
+				ds.Spec.UpdateStrategy = *strategy
+				_, err := dsClient.Create(context.TODO(), ds, metav1.CreateOptions{})
+				if err != nil {
+					t.Fatalf("Failed to create DaemonSet: %v", err)
+				}
+				defer cleanupDaemonSets(t, clientset, ds)
+
+				addNodes(nodeClient, 0, 1, nil, t)
+
+				validateDaemonSetPodsAndMarkReady(podClient, podInformer, 1, t)
+				validateDaemonSetStatus(dsClient, ds.Name, 1, t)
+				podsListBefore := podInformer.GetIndexer().List()
+				pod0Before := podsListBefore[0].(*v1.Pod)
```

Comment:
You'll need to `podsListBefore[0].(*v1.Pod).DeepCopy()` the pod, otherwise the mutation detector will complain that the pod in cache has been modified. 

But the test is very good proof, I'm happy it worked this well :)

Reply:
This is the root cause of the ingration failure on this PR.

Reply:
Thanks for pointing this out, fixed now.